ssh/tailssh/tailssh.go

```diff
@@ -254,141 +254,144 @@ func (c *conn) vlogf(format string, args ...any) {
-}
+	c.action0 = action
 
-// doPolicyAuth verifies that conn can proceed.
-// It returns nil if the matching policy action is Accept or
-// HoldAndDelegate. Otherwise, it returns errDenied.
-func (c *conn) doPolicyAuth(ctx ssh.Context) error {
-	if err := c.setInfo(ctx); err != nil {
-		c.logf("failed to get conninfo: %v", err)
-		return errDenied
-	}
-	a, localUser, acceptEnv, err := c.evaluatePolicy()
-	if err != nil {
-		return fmt.Errorf("%w: %v", errDenied, err)
-	}
-	c.action0 = a
-	c.currentAction = a
-	c.acceptEnv = acceptEnv
-	if a.Message != "" {
-		if err := ctx.SendAuthBanner(a.Message); err != nil {
-			return fmt.Errorf("SendBanner: %w", err)
-		}
-	}
-	if a.Accept || a.HoldAndDelegate != "" {
-		if a.Accept {
-			c.finalAction = a
-		}
+	if action.Accept || action.HoldAndDelegate != "" {
+		// Immediately look up user information for purposes of generating
+		// hold and delegate URL (if necessary).
 		lu, err := userLookup(localUser)
 		if err != nil {
-			c.logf("failed to look up %v: %v", localUser, err)
-			ctx.SendAuthBanner(fmt.Sprintf("failed to look up %v\r\n", localUser))
-			return err
+			err := fmt.Errorf("failed to look up %v: %w", localUser, err)
+			return nil, bannerError(err.Error(), err)
 		}
 		gids, err := lu.GroupIds()
 		if err != nil {
-			c.logf("failed to look up local user's group IDs: %v", err)
-			return err
+			err := fmt.Errorf("failed to look up local user's group IDs: %w", err)
+			c.logf(err.Error())
+			return nil, err
 		}
 		c.userGroupIDs = gids
 		c.localUser = lu
-		return nil
+		c.acceptEnv = acceptEnv
 	}
-	if a.Reject {
-		c.finalAction = a
-		return errDenied
+
+	for {
+		switch {
+		case action.Accept:
+			metricTerminalAccept.Add(1)
+			if action.Message != "" {
+				if err := c.spac.SendAuthBanner(action.Message); err != nil {
+					return nil, fmt.Errorf("error sending auth welcome message: %w", err)
+				}
+			}
+			c.finalAction = action
+			return &gossh.Permissions{}, nil
+		case action.Reject:
+			metricTerminalReject.Add(1)
+			c.finalAction = action
+			return nil, errDenied(action.Message)
+		case action.HoldAndDelegate != "":
+			if action.Message != "" {
+				if err := c.spac.SendAuthBanner(action.Message); err != nil {
+					return nil, fmt.Errorf("error sending hold and delegate message: %w", err)
+				}
+			}
+
+			url := action.HoldAndDelegate
+			if url == "" {
+				metricTerminalMalformed.Add(1)
+				return nil, errors.New("reached Action that lacked Accept, Reject, and HoldAndDelegate")
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Minute)
+			defer cancel()
+
+			metricHolds.Add(1)
+			url = c.expandDelegateURLLocked(url)
+
+			var err error
+			action, err = c.fetchSSHAction(ctx, url)
+			if err != nil {
+				metricTerminalFetchError.Add(1)
+				return nil, fmt.Errorf("fetching SSHAction from %s: %w", url, err)
+			}
+		default:
+			metricTerminalMalformed.Add(1)
+			return nil, errors.New("final action was neither accept nor reject")
```

Comment:
can we make these as banner error too?

Reply:
Same question, we don't currently show a banner for this condition.